lib/KarmaServer.js

```diff
@@ -0,0 +1,170 @@
+
+    runner = require('karma').runner;
+}
+
+/**
+ * Set the status of the server instance to the given status.
+ *
+ * @param status {number} The new status the server instance should have
+ * @private
+ */
+KarmaServer.prototype._setStatus = function(status) {
+    this._status = status;
+};
+
+/**
+ * Start the Karma server instance, if it has not been started before
+ *
+ * @returns {*|promise} a promise that will resolve with the instance itself when the instance starts properly within
+ *                      config.waitForServerTime seconds, and reject otherwise
+ */
+KarmaServer.prototype.start = function() {
+    var self = this,
+        deferred = Q.defer(),
+        startTime = Date.now(),
+        serverTimeout;
+
+    // Only servers that have not yet been started can be started
+    if(self._status === null) {
+        console.log('Starting a Karma server on port ' + self._config.port + '...');
+        self._setStatus(KarmaServerStatus.INITIALIZING);
+
+        // Limit the time it can take for a server to start to config.waitForServerTime seconds
+        serverTimeout = setTimeout(function() {
+            self._setStatus(KarmaServerStatus.DEFUNCT);
+            deferred.reject(
+                'Could not connect to a Karma server on port ' + self._config.port + ' within ' +
+                self._config.waitForServerTime + ' seconds'
+            );
+        }, self._config.waitForServerTime * 1000);
+
+        // Start a new Karma server process
+        self._serverProcess = spawn(
+            'node',
+            [
+                path.join(__dirname, 'run-karma-in-background.js'),
+                JSON.stringify(self._config)
+            ],
+            { stdio: 'pipe' }
+        );
+
+        // Handle data the Karma server writes to stdout
+        self._serverProcess.stdout.on('data', function(data) {
+            var message = data.toString('utf-8');
+
+            // This is a hack: Karma exposes no method of determining when the server is started up. We dissect the logs
+            if(message.indexOf('Connected on socket') > -1) {
+                clearTimeout(serverTimeout);
+                self._setStatus(KarmaServerStatus.READY);
+                console.log(
+                    'Server started after ' + (Date.now() - startTime) + 'ms and is listening on port ' +
+                    self._config.port
+                );
+                deferred.resolve(self);
+            }
+        });
+    } else {
+        deferred.reject('Server has already been started');
+    }
+
+    return deferred.promise;
+};
+
+/**
+ * Run the Karma tests on the server instance, if the instance is ready to run tests
+ *
+ * @returns {*|promise} a promise that will resolve with the test result when no errors occur and the run does not
+ *                      exceed config.waitForRunnerTime seconds, and reject otherwise
+ */
+KarmaServer.prototype.runTests = function() {
```

Comment:
Not sure whether this method belongs in this file, being a "KarmaServer" manager.